xds/internal/balancer/cdsbalancer/cdsbalancer.go

```diff
@@ -365,3 +502,39 @@ func (b *cdsBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.Sub
 func (b *cdsBalancer) Close() {
 	b.closed.Fire()
 }
+
+// ccWrapper wraps the balancer.ClientConn that was passed in to the CDS
+// balancer during creation and intercepts the NewSubConn() call from the child
+// policy. Other methods of the balancer.ClientConn interface are not overridden
+// and hence get the original implementation.
+type ccWrapper struct {
+	balancer.ClientConn
+
+	mu    sync.Mutex
+	xdsHI *xds.HandshakeInfo
+}
+
+// NewSubConn intercepts NewSubConn() calls from the child policy and adds an
+// address attribute which provides all information required by the xdsCreds
+// handshaker to perform the TLS handshake.
+func (ccw *ccWrapper) NewSubConn(addrs []resolver.Address, opts balancer.NewSubConnOptions) (balancer.SubConn, error) {
+	ccw.mu.Lock()
+	defer ccw.mu.Unlock()
+
+	if ccw.xdsHI == nil {
+		return ccw.ClientConn.NewSubConn(addrs, opts)
+	}
+
+	newAddrs := make([]resolver.Address, len(addrs))
+	for i, addr := range addrs {
+		newAddrs[i] = xds.SetHandshakeInfo(addr, ccw.xdsHI)
+	}
+	return ccw.ClientConn.NewSubConn(newAddrs, opts)
+}
+
+// setHandshakeInfo updates the handshakeInfo within the wrapper.
+func (ccw *ccWrapper) setHandshakeInfo(hi *xds.HandshakeInfo) {
+	ccw.mu.Lock()
```

Comment:
Related questions:

If the handshake info is updated, should we update the addresses in all the subconns?

Do we need to intercept future `SubConn.UpdateAddresses` as well?

Reply:
Existing addresses have an attribute which stores a pointer to the `xds.HandshakeInfo`. So, when we update the handshake info object, they will automatically pick it when on their next retry attempt.

We haven't done much so far with intercepting `SubConn.UpdateAddresses` in xds code so far. But a quick look does indicate that we might have to look into it. Added a TODO for now.